DataSource/Util/ChoicesBuilder.php

```diff
@@ -32,23 +32,31 @@ public function build($choicesConfiguration)
 
         if (
             is_array($choicesConfiguration) &&
-            isset($choicesConfiguration['repository']) &&
-            isset($choicesConfiguration['field'])
+            isset($choicesConfiguration['repository'])
         ) {
             if ($choicesConfiguration['repository'] instanceof EntityRepository) {
                 $repository = $this->getSpecifiedRepository($choicesConfiguration['repository']);
             } else {
                 $repository = $this->entityManager->getRepository($choicesConfiguration['repository']);
             }
-            $choicesField = $choicesConfiguration['field'];
-            $choicesSort = isset($choicesConfiguration['sort']) ? $choicesConfiguration['sort'] : null;
-            $choices = $this->getChoicesFromRepository($repository, $choicesField, $choicesSort);
+
+            if (isset($choicesConfiguration['field'])) {
+                $choices = $this->getChoicesFromRepositoryForField(
+                    $repository,
+                    $choicesConfiguration['field'],
+                    isset($choicesConfiguration['sort']) ? $choicesConfiguration['sort'] : null
+                );
+            } elseif (isset($choicesConfiguration['method'])) {
+                $choices = $this->getChoicesFromRepositoryWithMethod(
+                    $repository,
+                    $choicesConfiguration['method']
+                );
+            } else {
+                throw new \Exception('Repository source expects field or method parameter');
+            }
         } elseif (is_callable($choicesConfiguration)) {
             // Choices is a callable. It will generate the choices array.
-            $choices = $choicesConfiguration();
-            if (!is_array($choices)) {
-                throw new \Exception("Choices callback defined in table configurations must return arrays");
-            }
+            $choices = $this->getChoicesFromCallable($choicesConfiguration);
```

Comment:
you can return early here. and it seems you dont need the comment above (it could turn into [basmati rice](https://mobile.twitter.com/nzkoz/status/538892801941848064/photo/1))